ewatercycle/models/marrmot.py

```diff
@@ -259,3 +259,258 @@ def parameters(self) -> Iterable[Tuple[str, Any]]:
+        super().__init__()
+        self.version = version
+        self._parameters = [1000.0, 0.5, 0.5, 100.0, 0.5, 4.25, 2.5]
+        self.store_ini = [900.0, 900.0]
+        self.solver = Solver()
+        self._check_forcing(forcing)
+
+        self._set_singularity_image()
+        self._set_docker_image()
+
+    def _set_docker_image(self):
+        images = {
+            '2020.11': 'ewatercycle/marrmot-grpc4bmi:2020.11'
+        }
+        self.docker_image = images[self.version]
+
+    def _set_singularity_image(self):
+        images = {
+            '2020.11': 'ewatercycle-marrmot-grpc4bmi_2020.11.sif'
+        }
+        if CFG.get('singularity_dir'):
+            self.singularity_image = CFG['singularity_dir'] / images[self.version]
+
+    # unable to subclass with more specialized arguments so ignore type
+    def setup(self,  # type: ignore
+              maximum_soil_moisture_storage: float = None,
+              threshold_flow_generation_evap_change: float = None,
+              leakage_saturated_zone_flow_coefficient: float = None,
+              zero_deficit_base_flow_speed: float = None,
+              baseflow_coefficient: float = None,
+              gamma_distribution_chi_parameter: float = None,
+              gamma_distribution_phi_parameter: float = None,
+              initial_upper_zone_storage: float = None,
+              initial_saturated_zone_storage: float = None,
+              start_time: str = None,
+              end_time: str = None,
+              solver: Solver = None,
+              work_dir: Path = None) -> Tuple[Path, Path]:
+        """Configure model run.
+
+        1. Creates config file and config directory based on the forcing variables and time range
+        2. Start bmi container and store as :py:attr:`bmi`
+
+        Args:
+            maximum_soil_moisture_storage: in mm. Range is specfied in `model parameter range file <https://github.com/wknoben/MARRMoT/blob/master/MARRMoT/Models/Parameter%20range%20files/m_01_collie1_1p_1s_parameter_ranges.m>`_.
+            threshold_flow_generation_evap_change.
+            leakage_saturated_zone_flow_coefficient: in mm/d.
+            zero_deficit_base_flow_speed: in mm/d.
+            baseflow_coefficient: in mm-1.
+            gamma_distribution_chi_parameter.
+            gamma_distribution_phi_parameter.
+            initial_upper_zone_storage: in mm.
+            initial_saturated_zone_storage: in mm.
+            start_time: Start time of model in UTC and ISO format string e.g. 'YYYY-MM-DDTHH:MM:SSZ'. If not given then forcing start time is used.
+            end_time: End time of model in  UTC and ISO format string e.g. 'YYYY-MM-DDTHH:MM:SSZ'. If not given then forcing end time is used.
+            solver: Solver settings
+            work_dir: a working directory given by user or created for user.
+        Returns:
+            Path to config file and path to config directory
+        """
+        arguments = vars()
```

Comment:
Did not know you could use `vars()` to get the arguments as a dict. Thanks for the lesson

Reply:
yes, this is one of the [built-in functions](https://docs.python.org/3/library/functions.html), here I think that I could use both [vars()](https://docs.python.org/3/library/functions.html#vars) and [locals()](https://docs.python.org/3/library/functions.html#locals). But I chose the former.